cpp/src/arrow/compute/kernels/scalar_cast_extension.cc

```diff
@@ -42,14 +42,17 @@ Status CastToExtension(KernelContext* ctx, const ExecSpan& batch, ExecResult* ou
 
 std::shared_ptr<CastFunction> GetCastToExtension(std::string name) {
   auto func = std::make_shared<CastFunction>(std::move(name), Type::EXTENSION);
-  for (auto types : {IntTypes(), FloatingPointTypes(), StringTypes(), BinaryTypes()}) {
+  for (auto types : {PrimitiveTypes(), IntervalTypes(), TemporalTypes()}) {
     for (auto in_ty : types) {
       DCHECK_OK(
           func->AddKernel(in_ty->id(), {in_ty}, kOutputTargetType, CastToExtension));
     }
   }
-  DCHECK_OK(func->AddKernel(Type::DICTIONARY, {InputType(Type::DICTIONARY)},
-                            kOutputTargetType, CastToExtension));
+  for (auto in_ty : {Type::DICTIONARY, Type::LIST, Type::LARGE_LIST,
+                     Type::FIXED_SIZE_LIST, Type::EXTENSION}) {
```

Comment:
And decimal?

Reply:
Ideally, we would have a single loop enumerating all type ids (except `Type::EXTENSION`).
Something like:
```c++
for (Type::type in_ty : AllTypeIds()) {
  if (in_ty != Type::EXTENSION) {
    ...
```

Reply:
Updated in https://github.com/apache/arrow/pull/14106/commits/29db94b04c8d5598002a4942088cd3a556ffae86. Let me know if the `AllTypeIds` should rather be in a new file `type_fwd.cc` or inlined as it is now in `type_fwd.h`. 

@pitrou in another comment it was determined it ought to be okay to cast from extension to extension. The implementation would first cast from the input Extension to the target Extension's storage type, so long as those are compatible, then to the final Extension type. Does that align with your view?

Reply:
> The implementation would first cast from the input Extension to the target Extension's storage type, so long as those are compatible, then to the final Extension type. 

Does it make sense? Let's say you have UUID which is backed by `fixed_size_binary(16)` and Complex128 which is also backed by `fixed_size_binary(16)`. You don't want to implicitly cast between each other.

Reply:
`AllTypeIds()` should be declared in `type_fwd.h` and defined in `type.cc`, IMHO.

Reply:
That might be ok as an unsafe cast of some sort? Though I guess I'd argue that in case you do want that, you should explicitly cast the storage array and not the extension array

Reply:
In the (albeit, contrived) test case of `TinyintType` -> `SmallintType` it is acceptable, and in your example not at all. Not sure where the potential errors in casting goes against principle of least surprise. :thinking: 

Would adding an option `allow_casting_between_extensions` be enough for users to opt-in to their chosen risk?

Reply:
That doesn't really sound useful for now. `TinyintType` and `SmallintType` are toy examples that do not reflect real-world use cases.

Reply:
In most cases, attempts at casing between unrelated extension types is a user error. Let's not let it pass silently.

Reply:
I am not sure we should special case extension->extension casts. For a given extension type, there might be casts based on the storage type that _also_ don't make much sense, so it is hard for us to know in general what cast makes sense or not. 

For a specific use case, in the geometry types, we might have an extension type backed by a nested type with float or double coordinate values. Being able to cast from Extension(List<float>) <-> Extension(List<double>) would make sense in our case.

Reply:
Well, we should be conservative and avoid making senseless things by default...

Being able to register extension-specific casts would be another separate possibility.

Reply:
Okay. Given what we discussed earlier; https://github.com/apache/arrow/pull/14106/commits/c1a90c857b8d1bdf0eaaeda71c12ce9db7e37719 makes it such that casting between extension types where storage types don't match will fail, while giving a tip on how to more explicitly accomplish the same objective.